File: Lesson 2 - Calculator (ObjC)/ObjCCalculator/CalculatorBrain.m
//
//  CalculatorBrain.m
//  ObjCCalculator
//
//  Created by Vince Li on 2016/7/28.
//  Copyright © 2016年 Vince. All rights reserved.
//

#import "CalculatorBrain.h"


@implementation PendingInfo
- (id)initWithArguments:(PendingOperation)pendingOperation firstOperand:(double)firstOperand{
    self = [super init];
    if(self) {
        _pendingOperation = pendingOperation;
        _firstOperand = firstOperand;
    }
    return self;
}
@end

@implementation TouchEvent
- (id)initWithArguments:(int)operationType symbol:(NSString*)symbol {
    self = [super init];
    if(self) {
        _operationType = operationType;
        _symbol = symbol;
    }
    return self;
}
@end



enum OperationTypes{
    Constant = 0,
    UnaryOperation,
    BinaryOperation,
    Equals,
    Digital,
    Back,
    Save,
    Restore
};

@interface CalculatorBrain ()
- (void) executeBinaryOperation;
@property NSDictionary* operationTypes;
@property NSDictionary* operations;
@property NSMutableArray* touchHistory;
@property NSMutableArray* pendingOperationsForRestore;
@property Boolean pendingForRestore;

- (void) back;
- (void) _performOperarion: (NSString*) symbol addToHistory:(Boolean)addToHistory;
@end

@implementation CalculatorBrain
@synthesize brainDescription = _brainDescription;

- (id)init
{
    self = [super init];
    if (self)
    {
        _result = 0.0;
        _brainDescription = nil;
        _pendingInfo = nil;
        _touchHistory = [[NSMutableArray alloc] init];
        _pendingOperationsForRestore = [[NSMutableArray alloc] init];
        _variableValues = [[NSMutableDictionary alloc] init];
        _pendingForRestore = false;
        
        _operationTypes = [NSDictionary dictionaryWithObjectsAndKeys:
                           [NSNumber numberWithInt:Digital], @"1",
                           [NSNumber numberWithInt:Digital], @"2",
                           [NSNumber numberWithInt:Digital], @"3",
                           [NSNumber numberWithInt:Digital], @"4",
                           [NSNumber numberWithInt:Digital], @"5",
                           [NSNumber numberWithInt:Digital], @"6",
                           [NSNumber numberWithInt:Digital], @"7",
                           [NSNumber numberWithInt:Digital], @"8",
                           [NSNumber numberWithInt:Digital], @"9",
                           [NSNumber numberWithInt:Digital], @"0",
                           [NSNumber numberWithInt:Digital], @".",
                           [NSNumber numberWithInt:Constant], @"π",
                           [NSNumber numberWithInt:Constant], @"e",
                           [NSNumber numberWithInt:Constant], @"C",
                           [NSNumber numberWithInt:UnaryOperation], @"√" ,
                           [NSNumber numberWithInt:UnaryOperation], @"log" ,
                           [NSNumber numberWithInt:UnaryOperation], @"cos" ,
                           [NSNumber numberWithInt:UnaryOperation], @"sin" ,
                           [NSNumber numberWithInt:UnaryOperation], @"tan" ,
                           [NSNumber numberWithInt:UnaryOperation], @"±" ,
                           [NSNumber numberWithInt:BinaryOperation], @"+",
                           [NSNumber numberWithInt:BinaryOperation], @"-",
                           [NSNumber numberWithInt:BinaryOperation], @"×",
                           [NSNumber numberWithInt:BinaryOperation], @"÷",
                           [NSNumber numberWithInt:Equals], @"=",
                           [NSNumber numberWithInt:Back], @"⬅︎",
                           [NSNumber numberWithInt:Restore], @"M",
                           nil];
        
        _operations = [NSDictionary dictionaryWithObjectsAndKeys:
                       ^() {return M_PI;}, @"π",
                       ^() {return M_E;}, @"e",
                       ^() {return 0.0;}, @"C",
                       ^(double operand) {return sqrt(operand);}, @"√" ,
                       ^(double operand) {return log10(operand);}, @"log" ,
                       ^(double operand) {return cos(operand);}, @"cos" ,
                       ^(double operand) {return sin(operand);}, @"sin" ,
                       ^(double operand) {return tan(operand);}, @"tan" ,
                       ^(double operand) {return -operand;}, @"±" ,
                       ^(double p1, double p2) { return p1 + p2; }, @"+",
                       ^(double p1, double p2) { return p1 - p2; }, @"-",
                       ^(double p1, double p2) { return p1 * p2; }, @"×",
                       ^(double p1, double p2) { return p1 / p2; }, @"÷",
                       nil];
    }
    return self;
}

- (void) setOperand: (NSString*) operand {
    _result = operand.doubleValue;
}

- (double) pushOperand: (NSString*) operand {
    [_touchHistory addObject:[[TouchEvent alloc] initWithArguments:Save symbol:operand]];
    return [self evaluate];
}

- (void) executeBinaryOperation {
    if(_pendingInfo) {
        if(_pendingInfo.pendingOperation) {
            _result = _pendingInfo.pendingOperation( _pendingInfo.firstOperand, _result);
        }
        _pendingInfo = nil;
    }
}


- (void) _performOperarion: (NSString*) symbol addToHistory:(Boolean)addToHistory {
    NSNumber* operationType = [ _operationTypes objectForKey: symbol];
    if(operationType){
        int opType = [operationType intValue];
        switch(opType) {
            case Constant:
            {
                double(^block)() = [_operations objectForKey:symbol];
                if(block) {
                    _result = block();
                    _pendingInfo = nil;
                }
                if(addToHistory) {
                    [_touchHistory addObject:[[TouchEvent alloc] initWithArguments:opType symbol:symbol]];
                }
                break;
            }
            case UnaryOperation:
            {
                if(_pendingForRestore == false) {
                    [self executeBinaryOperation];
                    double(^block)(double) = [_operations objectForKey:symbol];
                    if(block) {
                        _result = block(_result);
                        _pendingInfo = nil;
                    }
                }
                else {
                    [_pendingOperationsForRestore addObject:[[TouchEvent alloc] initWithArguments:opType symbol:symbol]];
                }
                if(addToHistory) {
                    [_touchHistory addObject:[[TouchEvent alloc] initWithArguments:opType symbol:symbol]];
                }
                break;
            }
            case BinaryOperation:
            {
                if(_pendingForRestore == false) {
                    [self executeBinaryOperation];
                    double(^block)(double, double) = [_operations objectForKey:symbol];
                    if(block) {
                        _pendingInfo = [[PendingInfo alloc] initWithArguments:block firstOperand:_result];
                    }
                }
                else {
                    [_pendingOperationsForRestore addObject:[[TouchEvent alloc] initWithArguments:opType symbol:symbol]];
                }
                if(addToHistory) {
                    [_touchHistory addObject:[[TouchEvent alloc] initWithArguments:opType symbol:symbol]];
                }
                break;
            }
            case Equals:
                if(_pendingForRestore == false) {
                    [self executeBinaryOperation];
                }
                else {
                    [_pendingOperationsForRestore addObject:[[TouchEvent alloc] initWithArguments:opType symbol:symbol]];
                }
                if(addToHistory) {
                    [_touchHistory addObject:[[TouchEvent alloc] initWithArguments:opType symbol:symbol]];
                }
                break;
            case Digital:
                if(addToHistory) {
                    [_touchHistory addObject:[[TouchEvent alloc] initWithArguments:opType symbol:symbol]];
                }
                break;
            case Back:
                [self back];
                break;
            case Restore:
                if(addToHistory) {
                    [_touchHistory addObject:[[TouchEvent alloc] initWithArguments:opType symbol:symbol]];
                }
                if ([_variableValues objectForKey:symbol] == nil) {
                    _pendingForRestore = true;
                    [_pendingOperationsForRestore addObject:[[TouchEvent alloc] initWithArguments:opType symbol:symbol]];
                }
                else {
                    [self evaluate];
                }
                break;
            case Save:
                break;
        }
    }
}

- (void) performOperarion: (NSString*) symbol {
    [self _performOperarion:symbol addToHistory:TRUE];
}


- (NSString*) generateDescription {
    NSMutableString* output = [NSMutableString stringWithFormat:@""];
    for (int i = 0; i < [_touchHistory count]; i++) {
        TouchEvent* event = [_touchHistory objectAtIndex:i];
        if(event) {
            switch(event.operationType) {
                case Constant:
                    if (output.length > 0) {
                        [output appendString:@", "];
                    }
                    [output appendString:event.symbol];
                    break;
                case UnaryOperation:
                    if (output.length > 0) {
                        NSUInteger index = 0;
                        NSRange rangeToComma = [output rangeOfString:@", " options:NSBackwardsSearch];
                        
                        if(rangeToComma.location != NSNotFound) {
                            index = rangeToComma.location;
                        }
                        NSString *preOutput = [output substringToIndex:index];
                        NSString *postOutput = [output substringFromIndex:index];
                        output = [NSMutableString stringWithFormat:@"%@%@(%@)", preOutput, event.symbol, postOutput];
                    }
                    break;
                case BinaryOperation:
                    [output appendString:event.symbol];
                    break;
                case Equals:
                    output = [NSMutableString stringWithFormat:@""];
                    break;
                case Digital:
                    [output appendString:event.symbol];
                    break;
                case Back:
                    break;
                case Save:
                    break;
                case Restore:
                    [output appendString:event.symbol];
                    break;
            }
        }
    }
    return [NSString stringWithString:output];
}


- (bool) isValidInput: (NSString*) input currentDisplay:(NSString*)currentDisplay {
    if(currentDisplay) {
        if([input isEqualToString:@"."]) {
            if([currentDisplay rangeOfString:@"."].location != NSNotFound) {
                return false;
            }
        }
        else if([input isEqualToString:@"0"]) {
            if([currentDisplay rangeOfString:@"0"].location == 0 && [currentDisplay rangeOfString:@"."].location == NSNotFound) {
                return false;
            }
        }
        else {
            NSNumber* operationType = [ _operationTypes objectForKey: input];
            if(operationType && [operationType intValue] == BinaryOperation) {
                NSNumber* lastOperationType = [_operationTypes objectForKey:[_touchHistory lastObject]];
                if(lastOperationType && [lastOperationType intValue] == BinaryOperation) {
                    return false;
                }
            }
        }
    }
    return true;
}

- (void) back {
    if (_touchHistory.count > 0) {
        [_touchHistory removeLastObject];
    }
}

- (double) evaluate {
    if(_pendingForRestore == true && _pendingOperationsForRestore.count > 0) {
        bool allVariableSet = true;
        for(NSUInteger i = 0; i < _pendingOperationsForRestore.count; i++) {
            TouchEvent* event = [_pendingOperationsForRestore objectAtIndex:i];
            switch (event.operationType) {
                case Restore:
                    if ([_variableValues objectForKey:event.symbol] == nil) {
                        allVariableSet = false;
                    }
                    break;
                default:
                    break;
            }
        }
        
        if(allVariableSet) {
            _pendingForRestore = false;
            for(NSUInteger i = 0; i < _pendingOperationsForRestore.count; i++) {
                TouchEvent* event = [_pendingOperationsForRestore objectAtIndex:i];
                switch (event.operationType) {
                    case Restore:
                    {
                        NSString* value = [_variableValues objectForKey:event.symbol];
                        if(value) {
                            _result = [value doubleValue];
                        }
                    }
                        break;
                    default:
                        [self _performOperarion:event.symbol addToHistory:false];
                        break;
                }
            }
            [_pendingOperationsForRestore removeAllObjects];
        }
    }
    
    return _result;
}


- (void) setBrainDescription:(NSString *)brainDescription {
}
- (NSString*)brainDescription {
    NSString* output = [self generateDescription];
    return (output && output.length > 0) ? output : @" ";
}

@end

File: Lesson 2 - Calculator (Swift)/SwiftCalculator/CalculatorBrain.swift
//
//  CalculatorBrain.swift
//  SwiftCalculator
//
//  Created by Vince Li on 2016/7/27.
//  Copyright © 2016年 Vince. All rights reserved.
//

import Foundation


class CalculatorBrain {
    
    init () {
        self.variableValues = Dictionary<String, Double>()
    }
    private enum Operation {
        case Constant(Double)
        case UnaryOperation(Double -> Double)
        case BinaryOperation((Double, Double) -> Double)
        case Equals
        case Digital(String)
        case Back
        case Save
        case Restore
    }
    
    private var operations : Dictionary<String,Operation> = [
        "1" : Operation.Digital("1"),
        "2" : Operation.Digital("2"),
        "3" : Operation.Digital("3"),
        "4" : Operation.Digital("4"),
        "5" : Operation.Digital("5"),
        "6" : Operation.Digital("6"),
        "7" : Operation.Digital("7"),
        "8" : Operation.Digital("8"),
        "9" : Operation.Digital("9"),
        "0" : Operation.Digital("0"),
        "." : Operation.Digital("."),
        "π" : Operation.Constant(M_PI),
        "e" : Operation.Constant(M_E),
        "C" : Operation.Constant(0),
        "√" : Operation.UnaryOperation(sqrt),
        "log" : Operation.UnaryOperation(log10),
        "cos" : Operation.UnaryOperation(cos),
        "sin" : Operation.UnaryOperation(sin),
        "tan" : Operation.UnaryOperation(tan),
        "±" : Operation.UnaryOperation({ -$0 }),
        "+" : Operation.BinaryOperation({ $0 + $1 }),
        "-" : Operation.BinaryOperation({ $0 - $1 }),
        "×" : Operation.BinaryOperation({ $0 * $1 }),
        "÷" : Operation.BinaryOperation({ $0 / $1 }),
        "=" : Operation.Equals,
        "⬅︎": Operation.Back,
        "M": Operation.Restore,
        ]
    
    private struct PendingBinaryOperationInfo {
        var pendingOperation : (Double, Double) -> Double
        var firstOperand : Double
    }
    private var pendingInfo : PendingBinaryOperationInfo?
    
    private func executePendingBinaryOperation() {
        if pendingInfo != nil {
            accumulator = pendingInfo!.pendingOperation(pendingInfo!.firstOperand, accumulator)
            pendingInfo = nil
        }
    }
    
    private var accumulator = 0.0
    private var touchHistories = [(symbol:String, operation:Operation)]()
    private var pendingForRestore = false
    private var pendingOperationsForRestore = [(symbol:String, operation:Operation)]()
    
    func setOperand(operand: Double?) {
        if let op = operand {
            accumulator = op
        }
    }
    
    func performOperarion(symbol: String) {
        _performOperarion(symbol, addToHistory:true)
    }
    
    private func _performOperarion(symbol: String, addToHistory:Bool) {
        if let operation = operations[symbol] {
            switch operation {
            case .Constant(let value):
                accumulator = value
                pendingInfo = nil
                if(addToHistory) {
                    touchHistories.append( (symbol, operation) )
                }
            case .UnaryOperation(let function):
                if pendingForRestore == false {
                    executePendingBinaryOperation()
                    accumulator = function(accumulator)
                    pendingInfo = nil
                }
                else {
                    pendingOperationsForRestore.append((symbol, operation))
                }
                if(addToHistory) {
                    touchHistories.append( (symbol, operation) )
                }
            case .BinaryOperation(let function):
                if pendingForRestore == false {
                    executePendingBinaryOperation()
                    pendingInfo = PendingBinaryOperationInfo(pendingOperation: function, firstOperand: accumulator)
                }
                else {
                    pendingOperationsForRestore.append((symbol, operation))
                }
                if(addToHistory) {
                    touchHistories.append( (symbol, operation) )
                }
            case .Equals:
                if pendingForRestore == false {
                    executePendingBinaryOperation()
                }
                else {
                    pendingOperationsForRestore.append((symbol, operation))
                }
                if(addToHistory) {
                    touchHistories.append( (symbol, operation) )
                }
            case .Digital:
                if(addToHistory) {
                    touchHistories.append( (symbol, operation) )
                }
            case .Back:
                back()
            case .Restore:
                if(addToHistory) {
                    touchHistories.append( (symbol, operation) )
                }
                if variableValues[symbol] == nil {
                    pendingForRestore = true
                    pendingOperationsForRestore.append((symbol, operation))
                }
                else {
                    evaluate()
                }
            case .Save:
                break
            }
        }
    }
    
    private func generateDescription() -> String {
        var output:String = ""
        for (newString, element) in touchHistories {
            switch element {
            case .Constant:
                if !output.isEmpty {
                    output += ", "
                }
                output += newString
            case .UnaryOperation:
                if !output.isEmpty {
                    var index = 0
                    if let rangeToComma = output.rangeOfString(", ", options: NSStringCompareOptions.BackwardsSearch) {
                        index = output.startIndex.distanceTo(rangeToComma.endIndex)
                    }
                    let preOutput = output.substringToIndex(output.startIndex.advancedBy(index))
                    let postOutput = output.substringFromIndex(output.startIndex.advancedBy(index))
                    output = preOutput + String(format: "%@(%@)", newString, postOutput)
                }
            case .BinaryOperation:
                output += newString
            case .Equals:
                output = ""
            case .Digital(let value):
                output += value
            case .Back:
                break
            case .Restore:
                output += newString
            case .Save:
                break
            }
        }
        return output
    }
    
    private func back() {
        if touchHistories.isEmpty == false {
            touchHistories.removeLast()
        }
    }
    
    func isValidInput(input: String, currentDisplay:String) -> Bool {
        switch input {
        case ".":
            if currentDisplay.rangeOfString(".") != nil {
                return false;
            }
        case "0":
            if let rangeToZero = currentDisplay.rangeOfString("0") {
                if currentDisplay.startIndex.distanceTo(rangeToZero.startIndex) == 0 && currentDisplay.rangeOfString(".") == nil {
                    return false;
                }
            }
        default:
            if let operation = operations[input] {
                switch operation {
                case .BinaryOperation:
                    if let lastTouch = touchHistories.last {
                        switch lastTouch.1 {
                        case .BinaryOperation:
                            return false
                        default:
                            break
                        }
                    }
                    break
                default:
                    break
                }
            }
            break
        }
        return true;
    }
    var result : Double {
        get {
            return accumulator
        }
    }
    var description : String {
        get {
            let output = generateDescription()
            return (output == "") ? " " : output
        }
    }
    
    func evaluate() -> Double? {
        if pendingForRestore == true && pendingOperationsForRestore.isEmpty == false {
            var allVariableSet = true
            for (symbol, operation) in pendingOperationsForRestore {
                switch operation {
                case .Restore:
                    if variableValues[symbol] == nil {
                        allVariableSet = false
                    }
                default:
                    break
                }
            }
            
            if allVariableSet == true {
                pendingForRestore = false
                
                for (symbol, operation) in pendingOperationsForRestore {
                    switch operation {
                    case .Restore:
                        accumulator = variableValues[symbol]!
                        break
                    default:
                        _performOperarion(symbol, addToHistory: false)
                        break
                    }
                }
                pendingOperationsForRestore.removeAll()
            }
        }
        
        
        return accumulator
    }
    func pushOperand(symbol: String) -> Double? {
        variableValues[symbol] = nil
        touchHistories.append( (symbol, Operation.Save) )
        return evaluate()
    }
    var variableValues: Dictionary<String,Double>
    
}